### PR TITLE
Fixes #8540

### LINKED ISSUE #8540: Save Chat Log freezes Dream Seeker
**Problem Description**: Whenever I try to save my chat log Dream Seeker freezes and no amount of time recovers it

**What did you expect to happen**: To get the option to save my chat log

**What happened instead**: Ded

**Why is this bad/What are the consequences**: Because I want to remember all the things I did as a fun pAI and for possible logging on the player side, say someone griefs you and the admins don't catch it

**Steps to reproduce the problem**: Go to Goonchat gear icon, click and go to Save chat log. Then reload Dream Seeker because it froze

**Possibly related stuff (which gamemode was it? What were you doing at the time? Was
anything else out of the ordinary happening?)**: Might be just an issue with my system or Internet speeds but if it can be replicated then maybe not. I even tried early at round start with little text to save and still got a freeze

### PR DESCRIPTION
**Pull request by IK3I**
Fixes #8540 

# Option 1
Also adds a separate iframe explaining how to save the output since IE8 lacks support for doing things nicely.

## The dialogue box:
Save Window

## The result:
Output

# Option 2 (current commit)
Node server does the chat log creation and dispersal. Requires a slightly modified haste-server on the system. Further development of this route is only going to happen with a solid yes.
Benefits are everything is squeaky clean and looks nice. Plus I could write a download prompt directly into the server code.
:cl:
fix: Saving a chat log no longer hangs you.
/:cl:

### CODE REVIEW

**Comment by SamHPurp** [comment]
solid work, mate

**Comment by taukausanake** [comment]
So just to make sure, this does two things: one, it fixes the issue with saving as the feature was intended; and two, it tells people with IE8 the pleb way to do it.
Because if doing it the pleb way (at least with my version of IE) you have to paste as raw text to see anything. I would wish I'd get all the colors and some formatting if pasted in MS Word but that was not the case. Know any reason why that might be? Otherwise I guess the copy/paste way _kind of_ works. Didn't copy everything I highlighted last time. Also for me at least hotkeys didn't work. Had to right-click. Confirm if I'm not the only one with that issue too?

**Comment by Tayyyyyyy** [comment]
@taukausanake if you follow @IK3I's steps and put it into an html, then open that html in your browser, then copy that web page and paste it into word, it should copy into word with all the formatting.

But if you want to copy into word, you can also just select all the text in the chat window and copy and paste it into word yourself, without the save log feature.

**Comment by IK3I** [comment]
pretty much it gives you raw html with embedded css so you gotta parse it externally since the parsed output carries even more problems than the unparsed version. six hours of trudging through IE8 JS taught me this well.

**Comment by taukausanake** [comment]
OH I GET IT NOW. So no matter what it returns that window to you and you just got to copy the xml into an html and then open that. That is SO much better

**Comment by IK3I** [comment]
That is exactly what it does. Spits out the code to make a webpage, you just have to tell your computer to parse it as such with the html extension.

**Comment by tigercat2000** [comment]
I still wonder if there's any way to trigger a file download within the embedded IE window goonchat uses.

**Comment by IK3I** [comment]
I tried for so long, nothing in js can open the a file gui in IE8. If you want to try setting up a python interface you might be able to prompt a download from that, but with just JS and HTML supported by IE8, it's just not possible. you can't even save raw page sources in a friendly manner.

**Comment by tigercat2000** [comment]
It might be possible to do serverside downloads - Make Goonchat send all of it's compressed HTML to the server, which can package it up into a file and make the user save it with browse_ftp. The only problem with that would be that IIRC there's size limits, and of course sanitizing data like that is a pain.

**Comment by IK3I** [comment]
even the html has to get sanitized to display in this but it would definitely have to get pushed elsewhere to have it save. Any Idea how big a chat log gets before it starts rolling?

**Comment by tigercat2000** [comment]
Not sure- it shouldn't ever be *really* big, since the HTML and CSS by their nature are going to just be text.

**Comment by IK3I** [comment]
looking through the dm docs, I would need to send the data to a temp doc on the server and then call ftp on the file. That would mean the clientside js would need to send the file along with their ckey, then the server would turn around and toss it back to the client. I'll need to look at the references to figure out how byond handles event listeners.

**Comment by IK3I** [comment]
Added the second option and the js needed to do it. The modified haste server wouldn't actually be too hard to implement, but I don't want to send the maintainers down a dependency rabbit hole without a solid yes.

File: goon/browserassets/js/browserOutput.js [hunk 1]
```diff
@@ -871,32 +871,73 @@ $(function() {
...
-	$('#saveLog').click(function(e) {
-		var saved = '';
+	$('#saveLog').click(function(e) {	//note that this will only work in browsers supporting CORS (IE 8+ and literally every alternative)
+		var makeHaste = function(txt) { //creates a haste and returns the url
+			if (window.XMLHttpRequest) {
+				hasteRequest = new XMLHttpRequest();
+			} else {
+				hasteRequest = new ActiveXObject("Microsoft.XMLHTTP");
+			}
+			hasteRequest.open('POST', 'http://localhost:7777/documents', true);
```
Comment (tigercat2000) [inline goon/browserassets/js/browserOutput.js]: Could you make the `localhost:7777` part of this a setting in `opts`? https://github.com/ParadiseSS13/Paradise/blob/master/goon/browserassets/js/browserOutput.js#L24

File: goon/browserassets/js/browserOutput.js [hunk 2]
```diff
@@ -871,32 +871,73 @@ $(function() {
...
+						var data = JSON.parse(hasteRequest.responseText);	
+						if(data.key) {
+							var d = new Date();
+							var timestamp = d.getFullYear()+'-'+(d.getMonth()+1)+'-'+d.getDate()+' '+
+											d.getHours()+':'+d.getMinutes()+':'+d.getSeconds();
+					
+							internalOutput('Chat log generated. Please click the link below to view the raw data.');
+							internalOutput('To save your log: Open it, right click anywhere on the page, then select Save As.<br> \
+											For best results, save the file as [name].html then open it in any browser.');
+							internalOutput('<a href="http://localhost:7777/raw/'+data.key+'">Chat Log: '+timestamp+'</a>');
```
Comment (tigercat2000) [inline goon/browserassets/js/browserOutput.js]: Same here

File: goon/code/datums/browserOutput.dm [hunk 3]
```diff
@@ -11,7 +11,7 @@ var/list/chatResources = list(
 	"goon/browserassets/css/fonts/PxPlus_IBM_MDA.ttf",
 	"goon/browserassets/css/font-awesome.css",
 	"goon/browserassets/css/browserOutput.css",
-	"goon/browserassets/json/unicode_9_annotations.json"
+	"goon/browserassets/json/unicode_9_annotations.json",
```
Comment (tigercat2000) [inline goon/code/datums/browserOutput.dm]: This comma is now unnecessary

File: goon/browserassets/js/browserOutput.js [hunk 4]
```diff
@@ -871,32 +871,73 @@ $(function() {
 		setCookie('pingdisabled', (opts.pingDisabled ? 'true' : 'false'), 365);
 	});
 
-	$('#saveLog').click(function(e) {
-		var saved = '';
+	$('#saveLog').click(function(e) {	//note that this will only work in browsers supporting CORS (IE 8+ and literally every alternative)
```
Comment (tigercat2000) [inline goon/browserassets/js/browserOutput.js]: If it wouldn't be too difficult - could you keep the changes you made originally, and make a `useHastebin` boolean in `opts`? So, if a server operator doesn't want to be bothered with setting up hastebin, chat log saving isn't completely disabled?

**Comment by IK3I** [comment]
I'll go ahead and make those changes.
Then I'll add the boolean to the config defaulting to no. 
As for the server, I'll put the server.js, config.js, and an install readme/scripts in the filesystem for goonchat. What OS is the server running?

**Comment by tigercat2000** [comment]
The webserver runs on linux- The hastebin server should probably be in the tools/ directory. @IK3I

**Comment by IK3I** [comment]
@tigercat, I think I should revert this to the original and we go with that for now as a basic update. I'll be making a more tailored version of the server in a package style to solve the dependency issue, but that will take some time to implement and I think having a slightly inconvenient solution is better than leaving something that crashes the client up any longer.

Thoughts?

**Comment by tigercat2000** [comment]
@IK3I Whoops missed your response
That's probably fine, you should do a proper revert commit instead of a reset, though, so the code stays intact in history.

**Comment by IK3I** [comment]
Revert Commit is done, I'll be working on getting the server packaged on my own repo, should have something ready reasonably soon, but I don't have a huge amount of free time to work on it during the weekdays.

**Pull request merged**